WordPress/openverse-catalog
# Make Finnish DAG dated

## Fixes


Fixes #621 and hopefully #665

## Description


Converts the Finnish Museum provider DAG to a dated DAG, using the `last_indexed` filter to get records that were updated since the last DagRun.

In production, the Finnish provider DAG is ingesting huge amounts of data and ultimately timing out. Converting the DAG to a _dated_ DAG reduces the amount of data being consumed at any one time, and we hoped this would resolve the issue on its own.

Unfortunately I discovered that even when converted to a daily DAG, we still encounter large batches of data. Perhaps unsurprisingly for a museum provider, it seems that data is indexed very sparsely and in large batches at once. That means the huge majority of days have no data, while a small number of days have _very large_ amounts. Moreover, these batches are large enough to be buggy on their own. I described my steps to investigate and reproduce this in [this comment](https://github.com/WordPress/openverse-catalog/pull/879#issuecomment-1333078808), but essentially: for large querysets, the API will continue to return fallacious data to us no matter how large a `page number` we request, apparently into perpetuity. As a result ingestion continues until it times out.

I addressed this by copying a strategy from our Flickr DAG: we now break the ingestion day up into 48 half-hour increments and ingest data one time slice at a time. This prevents us from ever ingesting too large a dataset at once.

In summary, this PR:
* converts the Finnish museum to a dated DAG
* further breaks ingestion into half-hour increments to avoid the buggy API behavior
* updates the `start_date` of the DAG to November 2015, which is the earliest data I could find for the API
* adds some extra logging

## Notes
* I chose not to try pulling out the time splitting code into a shared utility, as this logic is currently _not working_ for Flickr. We may revisit it as part of that work
* This DAG would be incredibly noisy if we tried to backfill it all the way to 2015, so we should NOT do so unless we come up with a solution for silencing the Slack notifications.
* Because we're filtering on `last_indexed`, we'll handle updates in the regular ingestion flow (ie, no need for a separate reingestion DAG!) The `catchup` feature will handle recovering skipped days if the DAG is ever turned off for any reason.

I'll do some more investigation to try to understand what the current status of Finnish Museum data is in our catalog. It's not immediately clear to me how much we need to backfill, and we'll want to be pretty targeted about it since this will be a very noisy DAG. If #880 is merged, we'll be able to manually run the DAG for specific dates 😃 

## Testing Instructions

`just test`, run the Finnish DAG locally.

I was able to reproduce the time-out bug locally by running the DAG for November 3rd, 2015, and then confirm that ingestion worked for that day after adding in the time-splitting logic. If you would like to confirm this too, here's how:
* Make sure you don't have an `ingestion_limit` set
* Delete any local DagRuns you may have of the Finnish DAG
* Enable the Finnish DAG. It should automatically start running at `November 1st, 2015`, so you'll see the first two DagRuns pass with zero records. Then it will begin the "buggy" day
* Let it run! I got `147,245` records in a little over 2 hours 🎉 

## Checklist


[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
